tools/utils.py

```diff
@@ -104,6 +108,72 @@ def __cmp(self, other: 'Version', comparator: T.Callable[[T.Any, T.Any], bool])
         # versions are equal, so compare revisions
         return comparator(self._r, other._r)
 
+class _JSONFile(abc.ABC):
+    FILENAME: str
+
+    def __init__(self, _: T.Any): ...
+
+    @classmethod
+    def load(cls) -> T.Self:
+        with open(cls.FILENAME, encoding='utf-8') as f:
+            return cls(json.load(f))
+
+    def encode(self) -> str:
+        return json.dumps(self, indent=2, sort_keys=True) + '\n'
+
+    def save(self) -> None:
+        with open(f'{self.FILENAME}.new', 'w', encoding='utf-8') as f:
+            f.write(self.encode())
+        os.rename(f'{self.FILENAME}.new', self.FILENAME)
+
+class ProjectReleases(T.TypedDict):
+    dependency_names: T.NotRequired[list[str]]
+    program_names: T.NotRequired[list[str]]
+    versions: list[str]
+
+class Releases(dict[str, ProjectReleases], _JSONFile):
+    FILENAME = 'releases.json'
+
+class ProjectCIConfig(T.TypedDict):
+    build_on: T.NotRequired[dict[str, bool]]
+    build_options: T.NotRequired[list[str]]
+    fatal_warnings: T.NotRequired[bool]
+    has_provides: T.NotRequired[bool]
+    skip_dependency_check: T.NotRequired[list[str]]
+    skip_program_check: T.NotRequired[list[str]]
+    skip_tests: T.NotRequired[bool]
+    test_options: T.NotRequired[list[str]]
+
+    alpine_packages: T.NotRequired[list[str]]
+    brew_packages: T.NotRequired[list[str]]
+    choco_packages: T.NotRequired[list[str]]
+    debian_packages: T.NotRequired[list[str]]
+    msys_packages: T.NotRequired[list[str]]
+    python_packages: T.NotRequired[list[str]]
+
+class CIConfig(dict[str, ProjectCIConfig], _JSONFile):
+    FILENAME = 'ci_config.json'
+
+    @property
+    def broken(self) -> list[str]:
+        system = platform.system().lower()
+        return T.cast('list[str]', self.get(f'broken_{system}', []))
+
+def wrap_path(name: str) -> Path:
+    return Path('subprojects', f'{name}.wrap')
+
+def read_wrap(name: str) -> configparser.ConfigParser:
+    config = configparser.ConfigParser(interpolation=None)
+    config.read(wrap_path(name), encoding='utf-8')
+    return config
+
+def write_wrap(path: Path, config: configparser.ConfigParser) -> None:
+    # configparser write() adds multiple trailing newlines, collapse them
+    buf = io.StringIO()
+    config.write(buf)
+    with path.open('w', encoding='utf-8') as f:
```

Comment:
```suggestion
    with open(path, 'w', encoding='utf-8') as f:
```

works too. :)

path.open is just:

```python
    def open(self, mode='r', buffering=-1, encoding=None,
             errors=None, newline=None):
        """
        Open the file pointed to by this path and return a file object, as
        the built-in open() function does.
        """
        if "b" not in mode:
            encoding = io.text_encoding(encoding)
        return io.open(self, mode, buffering, encoding, errors, newline)
```

Reply:
Oh indeed.  I find `path.open()` a bit cleaner when I have a `Path`.  Not strongly tied to it though.